File: python/Q647_palindromic_substrings.py
#
# @lc app=leetcode id=647 lang=python3
#
# [647] Palindromic Substrings
#

# @lc code=start
class Solution:
    def countSubstrings(self, s: str) -> int:
        t: str = "@#" + "#".join(s) + "#$"
        p: list[int] = [0] * len(t)

        center: int = 0
        right: int = 0

        for i in range(1, len(t) - 1):
            if i <= right:
                p[i] = right - i if right - i < p[2 * center - i] else p[2 * center - i]

            while t[i - p[i] - 1] == t[i + p[i] + 1]:
                p[i] += 1

            if i + p[i] > right:
                right = i + p[i]
                center = i

        return sum(((x + 1) // 2 for x in p))


# @lc code=end

File: rust/src/leetcode/Q647_palindromic_substrings.rs
/*
 * @lc app=leetcode id=647 lang=rust
 *
 * [647] Palindromic Substrings
 */

// @lc code=start
impl Solution {
    pub fn count_substrings(s: String) -> i32 {
        let mut result: i32 = 0;
        let word: Vec<char> = s.chars().collect();
        let mut dp: Vec<Vec<bool>> = vec![vec![false; word.len()]; word.len()];

        for i in (0..word.len()).rev() {
            for j in i..word.len() {
                if word[i] == word[j] && (j - i <= 2 || dp[i + 1][j - 1]) {
                    dp[i][j] = true;
                    result += 1;
                }
            }
        }

        result
    }
}
// @lc code=end
